src/interfaces/cloud.rs

```diff
@@ -0,0 +1,410 @@
+    interfaces::{
+        generic_parameters::{
+            DemScheme, KmacKey, BLOCK_LENGTH, DEM_KEY_LENGTH, KMAC_KEY_LENGTH, KWI_LENGTH,
+            MASTER_KEY_LENGTH, TABLE_WIDTH, UID_LENGTH,
+        },
+        ser_de::serialize_set,
+    },
+    kmac,
+};
+
+pub(crate) struct FindexCloud {
+    pub(crate) token: Token,
+    pub(crate) base_url: Option<String>,
+}
+
+/// See `Token@index_id`
+pub const INDEX_ID_LENGTH: usize = 5;
+
+/// The callback signature is a kmac of the body of the request used to do
+/// authorization (checking if this client can call this callback)
+pub const CALLBACK_SIGNATURE_LENGTH: usize = 32;
+
+/// This key is used to derive a new 32 bytes Kmac key.
+pub const SIGNATURE_KEY_LENGTH: usize = 16;
+
+pub const FINDEX_CLOUD_DEFAULT_DOMAIN: &str = "https://findex.cosmian.com";
+
+/// Findex Cloud tokens are a string containing all information required to do
+/// requests to Findex Cloud (except the label because it is a value changing a
+/// lot).
+///
+/// The string is encoded as follow:
+/// 1. `index_id` `INDEX_ID_LENGTH` chars (see `Token@index_id`)
+/// 2. base64 representation of the different keys:
+///     1. `SIGNATURE_KEY_LENGTH` bytes of findex master key (this key is never
+/// sent to the Findex Cloud backend)
+///     2. 1 byte prefix identifying the next key
+///     3. `SIGNATURE_KEY_LENGTH` bytes of callback signature key
+///     4. 1 byte prefix identifying the next key
+///     5. …
+///
+/// Currently each callback have an associated signature key used in a kmac to
+/// send request to the backend. These key are only used for authorization
+/// and do not secure the index (the findex master key do). In the future, we
+/// could do optimization to avoid having one key for each callback but we want
+/// to disallow the server to differentiate a `fetch_entries` for a search or a
+/// `fetch_entries` for an upsert while still allowing fine grain permissions.
+pub(crate) struct Token {
+    /// This ID identify an index inside the Findex Cloud backend
+    /// This number is not sensitive, it's only an ID. If someone find this ID,
+    /// it cannot do requests on the index because it doesn't have the keys.
+    /// We do not use auto-increment integer ID because we don't want to leak
+    /// the number of indexes inside our database.
+    /// We do not use UUID because the token is limited in space.
+    /// The abritrary chosen length is `INDEX_ID_LENGTH`.
+    index_id: String,
+
+    pub(crate) findex_master_key: KeyingMaterial<MASTER_KEY_LENGTH>,
+
+    fetch_entries_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+    fetch_chains_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+    upsert_entries_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+    insert_chains_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+}
+
+impl Display for Token {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut keys = self.findex_master_key.to_vec();
```

Comment:
Since we know how many bytes we are going to write, you could allocate the required space at once.

Reply:
It is a lot of code, do you have an idea how to simplify it?

```
        let size = self.findex_master_key.len()
            + self
                .fetch_entries_key
                .as_ref()
                .map(|key| key.len() + 1)
                .unwrap_or(0)
            + self
                .fetch_chains_key
                .as_ref()
                .map(|key| key.len() + 1)
                .unwrap_or(0)
            + self
                .upsert_entries_key
                .as_ref()
                .map(|key| key.len() + 1)
                .unwrap_or(0)
            + self
                .insert_chains_key
                .as_ref()
                .map(|key| key.len() + 1)
                .unwrap_or(0);
```